MSC3006: Bot Interactions (#3006)
matrix-org/matrix-spec-proposals: proposals/3006-bot-interactions.md

# MSC3006: Bot Interactions

> Bots are currently only possible in irc style. They are entirely text based and not interactive like for example Telegram.

Reviewer: ```suggestion
Bots are currently only possible in IRC style. They are entirely text based and not interactive like, for example, Telegram bots.
```
Telegram is interactive? or the bots?

Author: Telegram bots were meant here :)

This is trying to solve that issue by using a similar backwards compatible way like the one in Telegram.

This is highly inspired by https://core.telegram.org/bots#keyboards and is trying to allow similar types 
of keyboards for bots to allow more "fancy" bots while keeping full backwards compatibility.

## Proposal

The proposal is to have multiple interaction events. Which tell the client what kind of interactions the 
bot has and the client replies with a specific format on that.

> _Note a deeper explanation of the events will be below the interaction flow_

Reviewer: ```suggestion
_Note: a deeper explanation of the events will be below the interaction flow_
```
Not absolutely necessary, but...

### Interaction Flow

1. The bot joins a room and sends a state event containing information about possible interactions:

```json
{
  "type": "m.bot.interactions",
  "state_key": "@faq_bot:example.matrix",

>   "command_prefix": "!",

Author: From @penn5 :
> Why does there have to be a command prefix? Why does it have to be the same for all commands?

The reason for the command prefix is to have it backwards compatible for cunstruction of the legacy message. From all bots I have seen so far in the matrix network this prefix always was the same across all commands the bot had. That is why I made this a extra key instead of having that in every interaction. Also doing it in the interaction would have added an additional "command" key to each interaction to allow for backwards compatibility.

Reviewer 2: Surely the current way, the command text is required to be equal to the name? Doesn't that mean that for backwards compatibility the name can't be human readable (or any more than the bot command used to be)? Surely it makes more sense to add a `text` key to each interaction to store the message that should be sent? Then, the bot is able to customise the fallback message...

Author: Hm yeah that also makes sense.

    {
      "type": "m.interaction.command",
      "description": "Shows the help of the FAQ Bot",
      "name": "help"
    },
    {
      "type": "m.interaction.button",
      "description": "Search question",
      "name": "Search"
    }
  ]
}
```

2. The clients shows a button to open the commands list and/or a button to open the custom keyboard

3. The user clicks on for example the help command one.

4. The client sends an event of type `m.room.message` and msgtype `m.text`. This allows us to have fallback for free while 
having typed commands when client and bot support it.

```json
{
  "content": {
    "body": "!help",
    "msgtype": "m.text",
    "m.interaction.command": {
      "command": "help",
      "target": "@faq_bot:example.matrix"
    }
  },
  "type": "m.room.message"
}
```

5. The Bot responds with a request for the client to display a separate custom keyboard for the next stage of 
this command

```json
{
  "content": {
    "event_id": "$random_id123",
    "interactions": [
      {
        "type": "m.interaction.button",
        "description": "The Search command",
        "name": "Search"
      }
    ]
  },
  "type": "m.interaction.stage.start"
}
```

> 6. The both now directly shows the new custom keyboard indicating the user currently is in a stage of the previous command.

Reviewer: ```suggestion
6. The bot now directly shows the new custom keyboard, indicating that the user currently is in a stage of the previous command.
```
Some more grammar nitpicking here.

Additionally a way of exiting that stage should be available.

7a. The user selects the exit way and the client now sends a stage end event and returns to normal chat operation.

```json
{
  "content": {
    "event_id": "$random_id123"
  },
  "type": "m.interaction.stage.end"
}
```

7b. If the user clicks the new button the client sends the command as in step 4 with additionally mentioning the `stage_id` under the 

> `m.inteaction.command` object. The bot after that can react either with a new stage or with the same `m.interaction.stage.end` 

Reviewer 3: ```suggestion
`m.interaction.command` object. The bot after that can react either with a new stage or with the same `m.interaction.stage.end` 
```

```

### Event Specification

#### m.bot.interactions

A state event defining the possible entry actions a bot has. It does not list any stage specific actions.
It is used to allow clients to display a custom keyboard and a list of command based on the defined data.

The `state_key` should be the mxid of the bot adding that event.

The `command_prefix` is used to support fallback for clients not supporting the MSC and is used to generate a irc style
bot message as fallback. (See "Fallback" section for Details).

The interactions array defines all supported entry actions of a bot, Those can be either of type: `m.interaction.command`
or `m.interaction.button`,

Example:

```json
{
  "type": "m.bot.interactions",
  "state_key": "@faq_bot:example.matrix",
  "command_prefix": "!",
  "interactions": [
    {
      "type": "m.interaction.command",
      "description": "Shows the help of the FAQ Bot",
      "name": "help"
    },
    {
      "type": "m.interaction.button",
      "description": "Search question",
      "name": "Search"
    }
  ]
}
```

#### Interaction event types

Both `m.interaction.command` and `m.interaction.button` are required to be an object consisting of:

> * It's `type`

Reviewer: ```suggestion
* Its `type`
```

This can happen either because of abusive bots or because someone else started a command. Possible ways to check this are:

* Track the commands locally in the client
* Check for a `m.room.message` event with `m.interaction.command` and the `sender` matching the logged in user.

Benefit of the second option is that this works across multiple used clients.

Example:

```json
{
  "content": {
    "event_id": "$random_id123",
    "interactions": [
      {
        "type": "m.interaction.button",
        "description": "The Search command",
        "name": "Search"
      }
    ]
  },
  "type": "m.interaction.stage.start"
}
```

#### `m.interaction.stage.end`

This event indicates a flow end.

The `event_id` is supposed to reference the event that was the start of the last event. It is not supposed to be the initial command.

This event can be sent by both the client as well as the bot. 
If a client sends the event the bot should think of this as if the user aborted the current action entirely.
If a bot sends the event the client should think of this as having completed a flow fully.

> This event should be ignored when the user not started a command or the user has not previously received a stage start event. 

Reviewer: ```suggestion
This event should be ignored when the user has not started a command or the user has not previously received a stage start event. 
```

This can happen either because of abusive bots or because someone else started a command. Possible ways to check this are:

* Track the commands locally in the client
* Check for a `m.room.message` event with `m.interaction.command` and the `sender` matching the logged in user.

Benefit of the second option is that this works across multiple used clients.

Example:

```json
{
  "content": {
    "event_id": "$random_id123"
  },
  "type": "m.interaction.stage.end"
}
```

## What is `m.interaction.button`?

> This type is meant to be like telegrams custom keyboard feature. It is used to display a button in the UI which behaves like a custom keyboard with fixed functions.

Reviewer: ```suggestion
This type is meant to be like Telegram's custom keyboard feature. It is used to display a button in the UI which behaves like a custom keyboard with fixed functions.
```

It is however in this MSC not meant to support inline buttons on events but those in theory can be done by reusing this type.

## How are stages meant to work

Imagine stages as submenus. They are indicating the next step in a greater action. 

If a stage start is sent the user moves into the next decision tree of a menu. If a stage end happens the user gets back to the top of this decision tree.
Stage end should not be treated as a back button.

## Potential issues

> * Priority of stages is not defined if 2 commands happen at the same time

Author: Timestamps are difficult across federated systems. However this most likely is already solved by the relations. (Would love to have some more input on this assumption from some that understand relations better than I do)

> * Priority of stages is not defined if 2 commands happen at the same time

Reviewer 3: > Timestamps are difficult across federated systems.

Read receipts only advance forwards, hence that order is always well defined and the problem boils down to retrieving the read receipts from the bot's server. And relations would not be useful across multiple unrelated commands from different users.

* Multiple users using multiple bots in a group is currently not handled directly. The MSC only demands to ignore stages if the user not started the command itself.

## Alternatives

> * MSC2929 - This seems to solve parts of this as well but doesn't seem to specify custom keyboards

Author: This proposal actually intentionally DOESNT allow complex flows. This is the result of the discussion that was in #matrix-client-developers:matrix.org . This is because at that point widgets are the correct way of doing this. Anything more complex than this most likely should be a widget as for client devs it gets increasingly difficult to add interactions more complex than this that also still have a uniform experience across all clients. 

TL;DR: This MSC intentionally limits itself to a small subset of interactions due to a discussion with various active client developers.

> # MSC3006: Bot Interactions

Author: > Thank you for the link. My main point was that buttons should be bound to a message. I agree that HTML is not ideal. I meant this to be more of a minimal example. The approach that every message can get a number of buttons in a special format sounds good to me.

Moving from https://github.com/matrix-org/matrix-doc/pull/3006#issuecomment-830638719 (cc @JoshuaDietz ) to a thread. (Next time please use thread to make the PR more readable please)

The goal of this MSC is actually to not do inline buttons. That would be a separate MSC and I do agree it should exist. This one is meant to be non message bound commands :) Thats why buttons are not bound to messages.

Reviewer 4: Yes I got that the goal of this MSC are not inline buttons. But as I described I don't really see how these keyboard buttons can work with matrix since every user can send them to every room as he wants (as long as there are no protections against that, i.e. a new permission). So if they are implemented in my opinion there needs to be a concept on how to avoid abuse of the feature (see the problems that I described in my previous comment)

Author: Well a) as "m.bot.interactions" is an own event you could restrict it by powerlevels b) thats more an overall "issue" in matrix if you want to call it like that. Because bots are not separate from users. They are the same thing. So solving this is imho out of scope of this MSC. This is only possible to solve by separating users and bots as otherwise it is impossible to make restrictions that work for bots but restrict users :/
